plugin/toerror/toerror.go

```diff
@@ -0,0 +1,106 @@
+//  Copyright 2017 Walter Schulze
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package curry contains the implementation of the curry plugin, which generates the deriveCurry function.
+//
+// The deriveCurry function curries the first two parameters of the input function.
+//   deriveCurry(f func(A, B, ...) T) func(A) func(B, ...) T
+package toerror
+
+import (
+	"fmt"
+	"go/types"
+
+	"github.com/awalterschulze/goderive/derive"
+)
+
+// NewPlugin creates a new curry plugin.
+// This function returns the plugin name, default prefix and a constructor for the curry code generator.
+func NewPlugin() derive.Plugin {
+	return derive.NewPlugin("toerror", "deriveToError", New)
+}
+
+// New is a constructor for the curry code generator.
+// This generator should be reconstructed for each package.
+func New(typesMap derive.TypesMap, p derive.Printer, deps map[string]derive.Dependency) derive.Generator {
+	return &gen{
+		TypesMap: typesMap,
+		printer:  p,
+	}
+}
+
+type gen struct {
+	derive.TypesMap
+	printer derive.Printer
+}
+
+func (g *gen) Add(name string, typs []types.Type) (string, error) {
+	if len(typs) != 2 {
+		return "", fmt.Errorf("%s does not have two argument. Expected (error, func (a) (b, bool))", name)
+	}
+	errTyp := typs[0]
+	if !derive.IsError(errTyp) {
+		return "", fmt.Errorf("First parameter should be of type error")
+	}
+	funcTyp := typs[1]
+	sig, ok := funcTyp.(*types.Signature)
+	if !ok {
+		return "", fmt.Errorf("%s, 2nd param %s, is not of type function", name, g.TypeString(typs[1]))
+	}
+	params := sig.Params()
+	if params.Len() != 1 {
```

Comment:
I don't know if the function really needs an argument.
We should be able to handle any amount of arguments, especially zero